File: ArrayAndHashing/TwoSum/twoSum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	l := make(map[int]int)
	for i, v := range nums {
		if t, ok := l[target-v]; ok {
			return []int{i, t}
		}
		l[v] = i
	}
	return []int{}
}

func main() {
	s1 := []int{2, 7, 11, 15}
	t1 := 9
	fmt.Printf("Test 1: %v\n", twoSum(s1, t1))

	s2 := []int{3, 2, 4}
	t2 := 6
	fmt.Printf("Test 2: %v\n", twoSum(s2, t2))

	s3 := []int{3, 3}
	t3 := 6
	fmt.Printf("Test 3: %v\n", twoSum(s3, t3))
}

File: ArrayAndHashing/TwoSum/twoSum.py
from typing import List


class Solution:
    # NOTE: a little cleaner
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        tracker = {}
        for i, n in enumerate(nums):
            diff = target - n
            if diff in tracker:
                return [tracker[diff], i]
            tracker[n] = i
        return []

    # NOTE: time = O(n)
    # def twoSum(self, nums: List[int], target: int) -> List[int]:
    #     tracker = {}
    #     final = []
    #     for i, n in enumerate(nums):
    #         j = tracker.get(target - n, None)
    #         if j != None:
    #             final.extend([j, i])
    #             break
    #         tracker[n] = i
    #     return final


if __name__ == "__main__":
    sol = Solution()

    s = [2, 7, 11, 15]
    t = 9
    r = sol.twoSum(s, t)
    print(f"T1: {r}")

    s = [4, 5, 6]
    t = 10
    r = sol.twoSum(s, t)
    print(f"T2: {r}")

    s = [5, 5]
    t = 10
    r = sol.twoSum(s, t)
    print(f"T3: {r}")
